### PR TITLE
Filter features implemented individually (helper functions)

### PR DESCRIPTION
**Pull request by anyachandorkar**
In this PR me and Ted made final edits to the filter verification methods in the parent Filter class and also implemented the actual filter in the child JSONFilter class. The implementation is broken up into multiple helper functions. There are 4 helper functions that individually target every feature: distance, rent, room type, availability and a couple more assistive functions to those which help reformat inputs to match comparison. Currently we have not accounted for default values or filtering with a few features not all.

### CODE REVIEW

File: BackEnd/app/filter/filter.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        for value in room_type.values():
+            if type(value)!=str:
+                return False
+        # asserting valid keys
+        if((room_type.get("Bedrooms") is None) or (room_type.get("Bathrooms") is None)):
+            return False
+
+        # assign correct counts to variables 
+        if(room_type["Bedrooms"][-1]=="+"):
+            bed_count = room_type["Bedrooms"][0:-1]
```
Comment (Yizong98) [inline BackEnd/app/filter/filter.py:38]: I think you can use room_type["Bedrooms"][:-1]

File: BackEnd/app/filter/filter.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if(room_type["Bathrooms"][-1]=="+"):
+            bath_count = room_type["Bathrooms"][0:-1]
+        else:
+            bath_count = room_type["Bathrooms"]
+
+        # asserting that all bedroom counts are positive ints 
+        if((not bed_count.isdigit())):
+            return False
+        # asserting that all bathroom counts are positive floats 
+        if((not bath_count.replace(".","",1).isdigit())):
```
Comment (Yizong98) [inline BackEnd/app/filter/filter.py:50]: well done

File: BackEnd/app/filter/filter.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+        return True
+
+    def verify_distance(self, distance: str) -> bool:
+        # asserting that input is given as 3 character string like "<20"
+        if(type(distance)!=str):
+            return False
+        try:
+            distance_op = distance[0]
+            time_per = distance[1:-5]
```
Comment (Yizong98) [inline BackEnd/app/filter/filter.py:61]: add a comment to explain the maximum value is 99999, five digit numbers

File: BackEnd/app/filter/filter.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            valid_operators = ["<", ">"]
+        except IndexError:
+            return False
+
+        # asserting that distance is given as numbers and operators are either > or <
+        if((not time_per.isdigit()) or (distance_op not in valid_operators)):
+            return False
+
+        # asserting that time is given as a positive number
+        if(int(time_per)<0):
```
Comment (Yizong98) [inline BackEnd/app/filter/filter.py:71]: isdigit already verifies it is a positive number. So It is fine to delete this check

File: BackEnd/app/filter/filter_json.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    files = client.list_objects_v2(Bucket=BUCKET,
+                                   Prefix=DIR)['Contents']
+    # read the data from the object
+    str_file = client.get_object(
+        Bucket=BUCKET, Key=files[-1]['Key'])['Body'].read().decode('UTF-8')
+    data = json.loads(str_file)
+    return data
+
+@attrs
+class JSONFilter(object):
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:46]: inherit the Filter! This is the exciting part.

File: BackEnd/app/filter/filter_json.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        arr_rent = []
+        if len(rent) == 2:
+            arr_rent.append(self.clean_rent(rent[0]))
+            arr_rent.append(self.clean_rent(rent[1]))
+        else:
+            arr_rent.append(self.clean_rent(rent[0]))
+            arr_rent.append(self.clean_rent(rent[0]))
+
+        # find the intersection of the criteria and arr_rent
+        intersection = range(max(arr_rent[0], criteria[0]), min(
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:78]: this is a mind-blowing technique. Bravo!

File: BackEnd/app/filter/filter_json.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
+from attr import attrs, attrib
+import boto3
+import json
+import os
+from app.util.util import *
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:5]: don't import *.  https://docs.google.com/document/d/1Gps_3J-Nu4ld0GVEEV3YXGUrykqKgmGivcOp7UWlTKI/edit#heading=h.litxuvobnld5

File: BackEnd/app/filter/filter_json.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        """
+        Check whether the roomType of the listing at index satisfy the criteria
+        """
+        criterias = {key: self.convert_room_type_input(val) for key, val in criterias.items()}
+        listing_room_type = self.process_room_type(self.data[index]['roomType'])
+        bool_bed = criterias['Bedroom'][1](
+            listing_room_type['Bedroom'], criterias['Bedroom'][0])
+        bool_bath = criterias['Bathroom'][1](
+            listing_room_type['Bathroom'], criterias['Bathroom'][0])
+        return bool_bed & bool_bath
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:114]: exactly what I meant. thanks!

File: BackEnd/app/filter/filter_json.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+    def convert_distance_input(self, input_str):
+        # converting string of "<20 mins" to tuple of num and operator
+        if(input_str[0] == '<'):
+            return (int(input_str[1:-5]), operator.le)
+        else:
+            return (int(input_str[1:-5]), operator.ge)
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:121]: same explanation needed here

File: BackEnd/app/filter/filter_json.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+    def filter_availability(self, criteria, index):
+        # converts both input and listing to datetime object and compares
+        criteria = self.convert_availability_input(criteria)
+        listing_availability = self.data[index]['availability'].lower()
+        if((listing_availability == "now") or (listing_availability == "today")):
+            return True
+        else:
+            listing_availability = datetime.strptime(listing_availability, "%m/%d/%Y")
+            if(listing_availability<criteria):
```
Comment (Yizong98) [inline BackEnd/app/filter/filter_json.py:145]: nested if elses are additional layers of abstractions. Think of different ways to make it more readable

Comment (Yizong98) [reply BackEnd/app/filter/filter_json.py:145]: bigger or equal is fine to address edge case. since it is latest acceptable

**Review by Yizong98 — CHANGES_REQUESTED** [review]
It is an amazing PR!! BRAVO! Just need some little tweaks to get approval. I can definitely see all the efforts put in here.

File: BackEnd/app/filter/filter.py (hunk 3 above)
Comment (tedfeng424) [reply BackEnd/app/filter/filter.py:61]: wait where does 99999 come from?

Comment (Yizong98) [reply BackEnd/app/filter/filter.py:61]: as in five digits

**Pull request closed without merging**